intermediate_source/rpc_param_server_tutorial.rst

```diff
@@ -0,0 +1,398 @@
+   # the RRef and passes along the given argument.
+   # Example: If the value held by the RRef is of type Foo, then
+   # remote_method(Foo.bar, rref, arg1, arg2) is equivalent to calling
+   # <foo_instance>.bar(arg1, arg2) on the remote node and getting the result
+   # back.
+
+   def remote_method(method, rref, *args, **kwargs):
+       args = [method, rref] + list(args)
+       return rpc.rpc_sync(rref.owner(), call_method, args=args, kwargs=kwargs)
+
+Now, we're ready to define our parameter server. We will subclass ``nn.Module`` and save a handle to our network defined above. We'll also save an input device which will be the device our input is transferred to before invoking the model.
+
+.. code-block:: python
+
+   # --------- Parameter Server --------------------
+   class ParameterServer(nn.Module):
+       def __init__(self, num_gpus=0):
+           super().__init__()
+           model = Net(num_gpus=num_gpus)
+           self.model = model
+           self.input_device = torch.device(
+               "cuda:0" if torch.cuda.is_available() and num_gpus > 0 else "cpu")
+
+Next, we'll define our forward pass. Note that regardless of the device of the model output, we move the output to CPU, as the Distributed RPC Framework currently only supports sending CPU tensors over RPC.
+
+.. code-block:: python
+
+   class ParameterServer(nn.Module):
+   ...
+       def forward(self, inp):
+               inp = inp.to(self.input_device)
+               out = self.model(inp)
+               # This output is forwarded over RPC, which as of 1.5.0 only accepts CPU tensors.
+               # Tensors must be moved in and out of GPU memory due to this.
+               out = out.to("cpu")
+               return out
+
+Next, we'll define a few miscellanous functions useful for training and verification purposes. The first, ``get_dist_gradients``\ , will take in a Distributed Autograd context ID and call into the ``dist_autograd.get_gradients`` API in order to retrieve gradients computed by distributed autograd. More information can be found in the `distributed autograd documentation <https://pytorch.org/docs/stable/rpc.html#distributed-autograd-framework>`_. Note that we also iterate through the resulting dictionary and convert each tensor to a CPU tensor, as the framework currently only supports sending tensors over RPC. Next, ``get_param_rrefs`` will iterate through our model parameters and wrap them as a (local) `RRef <https://pytorch.org/docs/stable/rpc.html#torch.distributed.rpc.RRef>`_. This method will be invoked over RPC by trainer nodes and will return a list of the parameters to be optimized on this node. This is required as input to the `Distributed Optimizer <https://pytorch.org/docs/stable/rpc.html#module-torch.distributed.optim>`_\ , which requires all parameters it must optimize as a list of ``RRef``\ s. 
+
+.. code-block:: python
+
+   # Use dist autograd to retrieve gradients accumulated for this model.
```

Comment:
is the mismatched indent intentional?

Reply:
It looks like this line is correct but the rest of the code block here is overindented. Fixing.